File: 3970-shortest-path-with-at-most-k-consecutive-identical-characters/3970-shortest-path-with-at-most-k-consecutive-identical-characters.java
import java.util.*;

class Solution {
    public int shortestPath(int n, int[][] edges, String labels, int k) {

        List<int[]>[] adj = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<>();
        }

        for (int[] e : edges) {
            int u = e[0];
            int v = e[1];
            int w = e[2];
            adj[u].add(new int[]{v, w});
        }

        long[][] dist = new long[n][k + 2];

        for (int i = 0; i < n; i++) {
            Arrays.fill(dist[i], Long.MAX_VALUE);
        }

        PriorityQueue<long[]> pq =
                new PriorityQueue<>((a, b) -> Long.compare(a[0], b[0]));

        dist[0][1] = 0;
        pq.offer(new long[]{0, 0, 1});

        while (!pq.isEmpty()) {
            long[] cur = pq.poll();

            long curd = cur[0];
            int curn = (int) cur[1];
            int curk = (int) cur[2];

            if (curd > dist[curn][curk]) {
                continue;
            }

            for (int[] edge : adj[curn]) {
                int ngh = edge[0];
                int wt = edge[1];

                int nextk;

                if (labels.charAt(curn) == labels.charAt(ngh)) {
                    nextk = curk + 1;
                } else {
                    nextk = 1;
                }

                if (nextk > k) {
                    continue;
                }

                long nd = curd + wt;

                if (nd < dist[ngh][nextk]) {
                    dist[ngh][nextk] = nd;
                    pq.offer(new long[]{nd, ngh, nextk});
                }
            }
        }

        long ans = Long.MAX_VALUE;

        for (int i = 1; i <= k; i++) {
            ans = Math.min(ans, dist[n - 1][i]);
        }

        return ans == Long.MAX_VALUE ? -1 : (int) ans;
    }
}

File: 3970-shortest-path-with-at-most-k-consecutive-identical-characters/3970-shortest-path-with-at-most-k-consecutive-identical-characters.cpp
class Solution {
public:
    int shortestPath(int n, vector<vector<int>>& edges, string labels, int k) {
        vector<pair<long long,long long>>adj[n]; 
        for(auto i:edges){
            long long u = i[0];
            long long v = i[1];
            long long w = i[2];
            adj[u].push_back({v,w});
        }
        vector<vector<long long>> dist(n,vector<long long>(k+2,LLONG_MAX));
        priority_queue<tuple<long long,long long,long long>,vector<tuple<long long,long long,long long>>,greater<tuple<long long,long long,long long>>> pq;
        dist[0][1]=0;
        pq.push({0,0,1});
        while(!pq.empty()){
            auto [curd,curn,curk] = pq.top();
            pq.pop();
            if(dist[curn][curk]<curd){
                continue;
            }
            for(auto i:adj[curn]){
                long long ngh = i.first;
                long long wt = i.second;
                long long nextk = curk;
                if(labels[curn]==labels[ngh]){
                    nextk+=1;
                }
                else{
                    nextk = 1;
                }
                if(curd+wt<dist[ngh][nextk] && nextk<=k){
                    dist[ngh][nextk]=wt+curd;
                    pq.push({dist[ngh][nextk],ngh,nextk});
                }

            }
        }
        long long ans = LLONG_MAX;
        for(int i=1;i<=k;i++){
            ans = min(ans,dist[n-1][i]);
        }
        return ans==LLONG_MAX?-1:ans;
        
    }
};